### PR TITLE
Fix RealmCollection.contains not respecting custom equal methods

### PR DESCRIPTION
**Pull request by cmelchior**
Got reported directly, and fixing it was easier than creating the issue :D

EDIT: `static.realm.io` is still insanely slow so CI times out on the unit tests. Will move into review once that is fixed

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,8 +2,10 @@
 
 ### Bug fixes
 
-* Fixed NPE problem happened in SharedRealm.finalize() (#3730).
+* NPE problem happened in SharedRealm.finalize() (#3730).
```
Comment (Zhuinden) [inline CHANGELOG.md]: are you sure you need `happened`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,9 +2,10 @@
 
 ### Bug fixes
 
-* Fixed NPE problem happened in SharedRealm.finalize() (#3730).
-* Fixed a build error when the project is using Kotlin (#4087).
-* Fixed a bug causing classes to be replaced by classes already in Gradle's classpath (#3568).
+* NPE problem in SharedRealm.finalize() (#3730).
+* `RealmList.contains()` and `RealmResults.contains()` now correctly uses custom `equals()` methods on Realm model classes.
```
Comment (kneth) [inline CHANGELOG.md]: "uses" -> "use"
"methods" -> "method"

File: realm/realm-library/src/androidTest/java/io/realm/RealmCollectionTests.java [hunk 3]
```diff
@@ -181,6 +214,18 @@ public void contains_null() {
...
 
+    // Test that the custom equal methods is being used when testing if an object is part of the
+    // collection
+    @Test
+    public void contains_customEqualMethod() {
+        RealmCollection<CustomMethods> collection = createCustomMethodsCollection(realm, collectionClass);
+        // This custom equals method will only consider the field `name` when comparing objects.
+        // So this unmanaged version should be equal to any object with the same value, managed
+        // or not.
+        CustomMethods unmanaged = new CustomMethods("Child0");
```
Comment (kneth) [inline realm/realm-library/src/androidTest/java/io/realm/RealmCollectionTests.java]: Only test first element? Maybe testing `Child + TEST_SIZE - 1` too (checking the iterators)?

Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmCollectionTests.java]: And how about testing if `Child + TEST_SIZE` is not contained?

**Comment by cmelchior** [comment]
Updated @kneth

File: realm/realm-library/src/androidTest/java/io/realm/entities/CustomMethods.java [hunk 4]
```diff
@@ -33,20 +42,20 @@ public void setName(String name) {
...
 
-    @Ignore
-    public boolean reverseEquals;
+    public RealmList<CustomMethods> getMethods() {
+        return methods;
+    }
 
     @Override
     public boolean equals(Object o) {
         if (!(o instanceof CustomMethods)) {
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/entities/CustomMethods.java]: - you're not checking for equal identity 
- passed object could be null

File: realm/realm-library/src/androidTest/java/io/realm/entities/CustomMethods.java [hunk 5]
```diff
@@ -33,20 +42,20 @@ public void setName(String name) {
...
         CustomMethods other = (CustomMethods) o;
-        if (isManaged() == other.isManaged() && other.name.equals(name)) {
-            return !reverseEquals;
-        } else {
-            return reverseEquals;
-        }
+
+        // Only compare name. Managed and unmanaged objects will be equal as long as they have the
+        // same value
+        return other.name.equals(name);
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/entities/CustomMethods.java]: - `other.name` is not required could be null

Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/entities/CustomMethods.java]: I suggest the following semi-generated impl
```java
@Override
public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || !(o instanceof CustomMethods)) return false;

    CustomMethods that = (CustomMethods) o;

// Only compare name. Managed and unmanaged objects will be equal as long as they have the same value
    return name != null ? name.equals(that.name) : that.name == null;
}
```

File: realm/realm-library/src/main/java/io/realm/RealmList.java [hunk 6]
```diff
@@ -720,19 +720,26 @@ public boolean load() {
...
-        boolean contains = false;
         if (managedMode) {
             realm.checkIfValid();
+
+            // Deleted objects can never be part of a RealmList
             if (object instanceof RealmObjectProxy) {
                 RealmObjectProxy proxy = (RealmObjectProxy) object;
-                if (proxy.realmGet$proxyState().getRow$realm() != null && realm.getPath().equals(proxy.realmGet$proxyState().getRealm$realm().getPath()) && proxy.realmGet$proxyState().getRow$realm() != InvalidRow.INSTANCE) {
-                    contains = view.contains(proxy.realmGet$proxyState().getRow$realm().getIndex());
+                if (proxy.realmGet$proxyState().getRow$realm() == InvalidRow.INSTANCE) {
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/RealmList.java]: why removing the comparison  of the path ? we need to assure we're comparing managed object from the same Realm?

File: realm/realm-library/src/main/java/io/realm/RealmResults.java [hunk 7]
```diff
@@ -189,14 +189,22 @@ public boolean isManaged() {
      */
     @Override
     public boolean contains(Object object) {
-        boolean contains = false;
-        if (isLoaded() && object instanceof RealmObjectProxy) {
-            RealmObjectProxy proxy = (RealmObjectProxy) object;
-            if (realm.getPath().equals(proxy.realmGet$proxyState().getRealm$realm().getPath()) && proxy.realmGet$proxyState().getRow$realm() != InvalidRow.INSTANCE) {
-                contains = (table.sourceRowIndex(proxy.realmGet$proxyState().getRow$realm().getIndex()) != TableOrView.NO_MATCH);
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/RealmResults.java]: we're not using `Java_io_realm_internal_TableView_nativeFindBySourceNdx` for table views? and instead doing a linear comparison with all `RealmResults` elements? what's the performance impact ?

Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmResults.java]: Probably quite big, but we are kinda forced to do that if you want to respect custom equal methods

File: realm/realm-library/src/main/java/io/realm/RealmList.java (hunk 6 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmList.java]: Our autogenerated `equals` method does that

File: realm/realm-library/src/androidTest/java/io/realm/entities/CustomMethods.java (hunk 4 above)
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/entities/CustomMethods.java]: Yeah, I might have missed a few nuances here. I'll double check

**Comment by cmelchior** [comment]
Updated @nhachicha

File: realm/realm-library/src/main/java/io/realm/RealmResults.java (hunk 7 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/RealmResults.java]: https://github.com/realm/realm-java/issues/4139 have been created to track a potential improvement to this.

**Pull request merged**